File: win-prog/ProgWinVB/Taming the Mouse/CaptureLoss/CaptureLoss.vb
'--------------------------------------------
' CaptureLoss.vb (c) 2002 by Charles Petzold
'--------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class CaptureLoss
    Inherits Form

    Shared Sub Main()
        Application.Run(new CaptureLoss())
    End Sub

    Sub New()
        Text = "Capture Loss"
    End Sub

    Protected Overrides Sub WndProc(ByRef msg As Message)
        ' Process WM_CAPTURECHANGED message.
        If msg.Msg = 533 Then Invalidate()
        MyBase.WndProc(msg)
    End Sub

    Protected Overrides Sub OnMouseDown(ByVal mea As MouseEventArgs)
        Invalidate()
    End Sub

    Protected Overrides Sub OnPaint(ByVal pea As PaintEventArgs)
        Dim grfx As Graphics = pea.Graphics
        If Capture Then
            grfx.FillRectangle(Brushes.Red, ClientRectangle)
        Else
            grfx.FillRectangle(Brushes.Gray, ClientRectangle)
        End If
    End Sub
End Class

File: win-prog/ProgWinCS/Taming the Mouse/CaptureLoss/CaptureLoss.cs
//------------------------------------------
// CaptureLoss.cs © 2001 by Charles Petzold
//------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class CaptureLoss: Form
{
     public static void Main()
     {
          Application.Run(new CaptureLoss());
     }
     public CaptureLoss()
     {
          Text = "Capture Loss";

               // Hook up NativeWindow object.

          CaptureLossWindow win = new CaptureLossWindow();
          win.form = this;
          win.AssignHandle(Handle);
     }
     protected override void OnMouseDown(MouseEventArgs mea)
     {
          Invalidate();
     }
     public void OnLostCapture()
     {
          Invalidate();
     }
     protected override void OnPaint(PaintEventArgs pea)
     {
          Graphics grfx = pea.Graphics;

          if (Capture)
               grfx.FillRectangle(Brushes.Red, ClientRectangle);
          else
               grfx.FillRectangle(Brushes.Gray, ClientRectangle);
     }
}

class CaptureLossWindow: NativeWindow 
{
     public CaptureLoss form;

     protected override void WndProc(ref Message message) 
     {
          if (message.Msg == 533)                 // WM_CAPTURECHANGED
               form.OnLostCapture();

          base.WndProc(ref message);
     }
}
